### ISSUE TITLE
New lint: `redundant_sort_by_key`.

### ISSUE THREAD

**Issue by pmnoxx** [rust-lang/rust-clippy#8212]
### What it does

I saw an interesting bug today. Apparently the following code is valid.

```rust
[3,1,2].sort_by_key(|x|{x;}|);
```
The extra semicolon makes this function return `()`, which is valid, but this makes the sorting pointless.

### Lint Name

`useless_sort_by_key`

### Category

correctness

### Advantage

- This is likely a bug, and should be fixed. I had it happen to my code, once.
- Or it can be just called redundant sort_by_key.

### Drawbacks

- I don't see any

### Example

```rust
[3,1,2].sort_by_key(|x|{x;}|);
```

Could be written as:

That codes does nothing can be removed.
Or can be fixed, but it may be impossible to figure out what the user intended
```rust
[3,1,2].sort_by_key(|x|{x}|);
```

I think this should be marked as an potential error.

**Comment by 5225225** [comment]
IMO this should apply to most/all `_by_key` methods.

So you'd be looking for all calls of a specific method with a signature like

```rust
fn max_by_key<B, F>(self, f: F) -> Option<Self::Item> where
    B: Ord,
    F: FnMut(&Self::Item) -> B, 
```

where `B` is `()`.

Though I suppose there may be legitimate use cases for a unit `_by_key` for some methods, so it might be a good idea to restrict it to just some types / be a bit smarter and let people explicitly return unit.
